Contributing guidelines (#251)
JuliaGaussianProcesses/KernelFunctions.jl: CONTRIBUTING.md

# Contribution guidelines

We follow the [ColPrac guide for collaborative practices](https://colprac.sciml.ai/). New contributors should make sure to read that guide.

## PRs and version numbers

> When contributing a PR, bump the version number (defined by `version = "..."` at the top of the base `Project.toml`) accordingly: generally this should be the patch version for backwards-compatible changes.

Reviewer: This is not true in general... Maybe just don't talk about the patch version at all since the different scenarios are discussed in the ColPrac guide?

Reviewer: ```suggestion
When contributing a PR, bump the version number (defined by `version = "..."` at the top of the base `Project.toml`) accordingly.
```

Author: rewrote this and linked to the corresponding ColPrac section

## Workflows

### How to make a new release (for organization members)

We use [JuliaRegistrator](https://github.com/JuliaRegistries/Registrator.jl#via-the-github-app):

> 1. Merge the bumped version number into `master` (see above); normally this will be the commit of a squash-merged pull request.
> 2. Go to the Github URL for this merge commit and write a comment saying `@JuliaRegistrator register`.

Reviewer: Technically, the merge process is separate from registering a new release with JuliaRegistrator:

```suggestion
On Github, go to the commit that you want to register and add a comment saying `@JuliaRegistrator register`.
```

Author: It may be technically different, but what's the process from a contributor's point of view? Do you want people to bump the version number & register a new release in each merge?

> The next release will then be processed automatically. KernelFunctions.jl does use TagBot so you can ignore the message about tagging. Note that it may take around 20 minutes until the actual release appears and you can edit the release notes if needed.

Reviewer: I just want to clarify (doesn't have to be added here, I think) that the Github releases and tags are nice but not relevant or mandatory for Julia packages (there exist many packages that work just fine without TagBot or Github releases). As soon as the PR by JuliaRegistrator is merged into the general registry, users can update KernelFunctions (well, in principle at least - sometimes it takes a while until the Julia PkgServers update the registry...).

Author: If not here, should this be clarified elsewhere? Again, this is about how would the process look like for a new contributor.

# Contribution guidelines

> We follow the [ColPrac guide for collaborative practices](https://colprac.sciml.ai/). New contributors should make sure to read that guide.

Reviewer 2: ```suggestion
We follow the [ColPrac guide for collaborative practices](https://colprac.sciml.ai/). New contributors should make sure to read the [first section](https://github.com/SciML/ColPrac#colprac-contributors-guide-on-collaborative-practices-for-community-packages) of that guide, but could also read the [Further Guidance](https://github.com/SciML/ColPrac#colprac-further-guidance) section if interested.
```
The whole guide is fairly long -- new contributors just need the highlights initially.

# Workflows

> ## Creating a new pull request

Reviewer 3: Should we add any more fundamental doc links like [this](https://docs.github.com/en/github/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-pull-requests) here or just assume any contributors know the basics?

Author: ```suggestion
## Creating a new [pull request](https://docs.github.com/en/github/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-pull-requests)
```

### Bumping the version number

> When contributing a PR, bump the version number (defined by `version = "..."` at the top of the base `Project.toml`) accordingly (as explained by the [guidance on the versioning scheme](https://colprac.sciml.ai/#incrementing-the-package-version) in the ColPrac guide).

Reviewer 2: ```suggestion
When contributing a PR, bump the version number (defined by `version = "..."` at the top of the base `Project.toml`) accordingly (as explained by the [guidance on the versioning scheme](https://colprac.sciml.ai/#incrementing-the-package-version) in the ColPrac guide).
If unsure about what the new version should be, please just open the PR anyway -- existing contrbutors will provide a suggestion.
```

### Running tests locally

> Run `make test`.

Reviewer 2: ```suggestion
Run `make test`.
These tests can take a long time to run, but there are a couple of things that can be done to alleviate this when improving a feature.
The first is to simply comment out the blocks of tests not required in `test/runtests.jl`.
Test files are paired 1-1 with source files, so if you're modifying code in `src/foo.jl`, you should only need to run the tests in `test/foo.jl` during development.
```
Should we also say something about `dev`ing the package?

Author: yeah that'd be good

Reviewer 2: Okay, how about:
```suggestion
Firstly, run
```julia
using Pkg
Pkg.develop("KernelFunctions")
```
at the Julia REPL, and navigate to `~/.julia/dev/KernelFunctions`.

Running `make test` with run the entire test suite.
These tests can take a long time to run, so it's often a good idea to simply comment out the blocks of tests not required in `test/runtests.jl`.
Test files are paired 1-1 with source files, so if you're modifying code in `src/foo.jl`, you should only need to run the tests in `test/foo.jl` during development.
```

# Contribution guidelines

We follow the [ColPrac guide for collaborative practices](https://colprac.sciml.ai/). New contributors should make sure to read the [first section](https://github.com/SciML/ColPrac#colprac-contributors-guide-on-collaborative-practices-for-community-packages) of that guide, but could also read the [Further Guidance](https://github.com/SciML/ColPrac#colprac-further-guidance) section if interested.

# Workflows

## Creating a new [pull request](https://docs.github.com/en/github/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-pull-requests)

### Bumping the version number

When contributing a PR, bump the version number (defined by `version = "..."` at the top of the base `Project.toml`) accordingly (as explained by the [guidance on the versioning scheme](https://colprac.sciml.ai/#incrementing-the-package-version) in the ColPrac guide).

> If unsure about what the new version should be, please just open the PR anyway -- existing contrbutors will provide a suggestion.

Reviewer 2: ```suggestion
If unsure about what the new version should be, please just open the PR anyway -- existing contributors will provide a suggestion.
```
my bad!
